docs/coding-and-design-guidelines.md

```diff
@@ -0,0 +1,50 @@
+
+## Prefer Microsoft abstractions
+
+Microsoft maintains a number of abstractions for common cross-cutting concerns in the .NET common library and in the `Microsoft.Extensions.*` namespaces. Where such an abstraction exists, we prefer to use it over an abstraction provided by a different assembly (including one of our own). 
+
+This makes it easier for us to isolate the application from 3rd party dependencies by relying on fairly stable abstractions. It also helps to keep parts of the application isolated from each other (i.e. being able to run the embedded database for maintenance mode without starting the NServiceBus endpoint).
+
+For example:
+
+- **Prefer IHostBuilder extensions over NServiceBus features**: Unless the code being added is specifically altering the NServiceBus endpoint, it makes sense to add new ServiceControl code directly as extensions to `IHostBuilder` rather than as NServiceBus `Feature` implementations. Some existing ServiceControl features may still be using the `Feature` abstraction to register its components. If these features are not modifying the NServiceBus endpoint, then they should be migrated to `IHostBuilder` extensions over time.
+- **Prefer Hosted Services over NServiceBus FeatureStartupTasks**: ServiceControl includes a lot of background tasks and these are both abstractions for building background tasks. `FeatureStartupTask` implementations are tied to the lifecycle of an endpoint. In general, we prefer to use the `IHostedService` abstraction which is tied to the lifecycle of the host application. NOTE: Hosted Services are started in the order that they are registered which allows us to better control the startup sequence. They are shut down in the reverse order.
+- **Prefer IServiceCollection over IConfigureComponents or IContainerBuilder**: Where possible we'd rather use the generic Microsoft DI abstractions over the NServiceBus one (IConfigureComponents) or the Autofac one (IContainerBuilder).
+  - When registering components from within an NServiceBus Feature, it still  makes sense to use IConfigureComponents but we should consider whether it makes sense to move the code out of NServiceBus features
+  - When relying on Autofac specific features, it makes sense to rely on IContainerBuilder. If a feature can be implemented in a way to avoid this, it should be preferred. We may in the future decide to remove our dependency on Autofac but that decision has not been made.
+- **Prefer IServiceProvider over ILifetimeService over IContainer**: As above, we'd rather rely on the generic Microsoft DI abstraction where possible and only fall back to NServiceBus or Autofac abstractions where strictly necessary.
+  - If relying on a specific feature of Autofac, then `ILifetimeService` should be used.
+  - `IContainer` should never be used. It is functionally equivalent to `IServiceProvider`
+
+### Exceptions
+
+There is one exception to the preference for Microsoft abstractions
```

Comment:
This doesn't sound like an exception to the preference, but rather a practical "it will take time to migrate to the preferred abstraction"

Reply:
Agreed, perhaps say "until NServiceBus core removes its static logging"?

Side note: this is tracked here https://github.com/Particular/PlatformArchitecture/blob/master/problems/nservicebus-core.md#custom-logging-abstraction

Reply:
We didn't want to make the decision that we are definitely moving away from NServiceBus logging, especially as it has not been deprecated for NServiceBus version 8. With that in mind, it is an exception to the preference. In this one case we do not prefer the Microsoft abstration.

@WilliamBZA @andreasohlund Do you think we need to make a stronger statement about it now?

Reply:
I'm not terribly concerned about it. If we haven't made a clear decision then that's fine too.